### PR TITLE
State and block execution

### LINKED ISSUE #65: Create Block Executor
Block Executor is a part of state package that is responsible for executing transactions from a block against application and maintaining the state (block hash, app hash, results hash, block height, etc).

### PR DESCRIPTION
**Pull request by tzdybal**
Resolves #65.

### CODE REVIEW

File: state/executor.go
Comment (liamsi) [inline state/executor.go:36]: We might want another name for this. Proposal is a consensus message (the thing that optimint does not do).

**Review by liamsi — COMMENTED** [review]
Nice to see this coming together 👍🏼 

Left some comments. Independent from my review, I think it would be good to have a second pair of eyes on this.

File: state/executor.go
Comment (evan-forbes) [inline state/executor.go:155]: perhaps we could have a `ValidateBasic` method for `types.Block` and call that here too. ref https://github.com/lazyledger/lazyledger-core/blob/8da16447f34680639033b531ed10b9ab756a14e0/types/block.go#L168-L168

File: state/executor.go
Comment (evan-forbes) [inline state/executor.go:189]: I think it would be a bonus if we could start handling all of these todo contexts. Would it make sense to use a context specific to the `*BlockExecutor` or to pass it as an argument?

**Review by evan-forbes — COMMENTED** [review]
I left a few more minor comments. I'm still getting better acquainted with optimint, but this seems like a great addition!

**Review by tzdybal — COMMENTED** [review]
OMG, just found old, mossy, open tab, with "Submit review" not clicked...

**Review by evan-forbes — APPROVED** [review]
I think makes sense to approve now, but ~~we should likely handle this comment before proceeding to merge https://github.com/celestiaorg/optimint/pull/58/files#r657950338~~ already discussed

Great work as per usual, Tomasz! I can't wait to play around with optimint more

:+1: :+1:

File: state/state.go
Comment (tzdybal) [reply state/state.go:7]: I would like to work on this (https://github.com/celestiaorg/optimint/issues/37) after MVP.

**Comment by tzdybal** [comment]
It's definitely not 100% done, but I'm closing PR as is. More changes and tests will be implemented with #82.

**Pull request merged**
